buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java

```diff
@@ -0,0 +1,115 @@
+ */
+package org.elasticsearch.gradle.plugin;
+
+import org.elasticsearch.gradle.VersionProperties;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.Project;
+import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Creates a plugin descriptor.
+ */
+public class PluginPropertiesTask extends DefaultTask {
+
+    private PluginPropertiesExtension extension;
+
+    private File descriptorOutput = new File(getProject().getBuildDir(),
+            "generated-resources/plugin-descriptor.properties");
+
+    public PluginPropertiesTask() {
+        Project project = this.getProject();
+        this.extension = project.getExtensions().create("esplugin", PluginPropertiesExtension.class, project);
+    }
+
+    @TaskAction
+    public void performAction() {
+        Project project = this.getProject();
+
+        File templateFile = new File(project.getBuildDir(), "templates/" + descriptorOutput.getName());
+        templateFile.getParentFile().mkdirs();
+
+        InputStream resourceTemplate = this.getClass().getResourceAsStream("/" + descriptorOutput.getName());
+
+        try {
+            Files.copy(resourceTemplate, templateFile.toPath());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        project.copy(copyArg -> {
+            copyArg.from(templateFile.getParentFile()).include(descriptorOutput.getName())
+                .into(descriptorOutput.getParentFile()).expand(this.generateSubstitutions());
+            this.getInputs().properties(this.generateSubstitutions());
+        });
+    }
+
+    private Map<String, String> generateSubstitutions() {
+        Map<String, String> substitutions = new HashMap<>();
+        substitutions.put("name", this.extension.getName());
+        substitutions.put("description", this.extension.getDescription());
+        substitutions.put("version", this.extension.getVersion());
+        substitutions.put("elasticsearchVersion", VersionProperties.getElasticsearch());
+        substitutions.put("javaVersion", getProject().getConvention().getPlugin(JavaPluginConvention.class)
+                .getTargetCompatibility().toString());
+        substitutions.put("classname", this.extension.getClassname());
+        substitutions.put("extendedPlugins", this.extension.getExtendedPlugins().stream().collect(Collectors.joining(",")));
+        substitutions.put("hasNativeController", String.valueOf(this.extension.hasNativeController()));
+        substitutions.put("requiresKeystore", String.valueOf(this.extension.isRequiresKeystore()));
+
+        return substitutions;
+    }
+
+    @OutputFile
```

Comment:
This annotation is incorrect here.
This could move into `generateSubstitutions` I think that's a proper place to do those checks.

Reply:
I will remote the `OutputFile` annotation. Why are we moving these `if-checks`?

Reply:
Without the annotations the method will never be called.